File: 323-NumberOfConnectedComponentsInAnUndirectedGraph/323-NumberOfConnectedComponentsInAnUndirectedGraph.py
# Last updated: 4/29/2026, 10:31:18 AM
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        from collections import defaultdict
        adj = defaultdict(set)
        
        for u, v in edges:
            adj[v].add(u)
            adj[u].add(v)
        
        count = 0
        visited = [False] * n
        
        def dfs(i):
            if not visited[i]:
                visited[i] = True
                for end in adj[i]:
                    dfs(end)
        
        for i in range(n):
            if not visited[i]:
                dfs(i)
                count += 1
        
        return count

File: 323-NumberOfConnectedComponentsInAnUndirectedGraph/323-NumberOfConnectedComponentsInAnUndirectedGraph.js
// Last updated: 4/29/2026, 10:31:23 AM
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
// 2022-12-08
// 05:25s
// O(e+v)
// O(e+v)
var countComponents = function(n, edges) {
    const map = {};
    for(const [u, v] of edges) {
        if(map[u] === undefined) {
            map[u] = new Set();
        }
        map[u].add(v);
        if(map[v] === undefined) {
            map[v] = new Set();
        }
        map[v].add(u);
        
    }
    console.log(map);
    const visited = new Array(n).fill(false);
    let count = 0;
    const dfs = start => {
        const ends = map[start];
        if(ends) {
            for(const end of ends) {
                if(!visited[end]) {
                    visited[end] = true;
                    dfs(end);
                }
            }
        }
    }
    for(let i = 0; i < n; i++) {
        if(!visited[i]) {
            visited[i] = true;
            count++;
            dfs(i);
        }
    }
    return count;
};